components/sidebar/sidebar_service.cc

```diff
@@ -225,75 +350,133 @@ SidebarService::ShowSidebarOption SidebarService::GetSidebarShowOption() const {
   DCHECK_NE(ShowSidebarOption::kShowOnClick, show_options);
   prefs_->SetInteger(kSidebarShowOption, static_cast<int>(show_options));
 }
 
 void SidebarService::LoadSidebarItems() {
-  auto* preference = prefs_->FindPreference(kSidebarItems);
-  if (preference->IsDefaultValue()) {
-    items_ = GetDefaultSidebarItems();
-    return;
-  }
-
-  const auto& items = preference->GetValue()->GetList();
-  for (const auto& item : items) {
-    SidebarItem::Type type;
-    if (const auto value = item.FindIntKey(kSidebarItemTypeKey)) {
-      type = static_cast<SidebarItem::Type>(*value);
-    } else {
-      continue;
-    }
+  auto default_items_to_add = GetDefaultSidebarItems();
 
-    std::string url;
-    if (const auto* value = item.FindStringKey(kSidebarItemURLKey)) {
-      url = *value;
-    } else {
-      continue;
-    }
-
-    // Always use latest properties for built-in type item.
-    if (type == SidebarItem::Type::kTypeBuiltIn) {
-      SidebarItem built_in_item;
-      if (const auto value = item.FindIntKey(kSidebarItemBuiltInItemTypeKey)) {
-        built_in_item = GetBuiltInItemForType(
-            static_cast<SidebarItem::BuiltInItemType>(*value));
+  // Pref for custom items and custom order
+  auto* preference = prefs_->FindPreference(kSidebarItems);
+  if (!preference->IsDefaultValue()) {
+    const auto& items = preference->GetValue()->GetList();
+    for (const auto& item : items) {
+      DVLOG(2) << "load: " << item.DebugString();
+      SidebarItem::Type type;
+      if (const auto type_value = item.FindIntKey(kSidebarItemTypeKey)) {
+        type = static_cast<SidebarItem::Type>(*type_value);
       } else {
-        // Fallback when built-in item type key is not existed.
-        built_in_item = GetBuiltInItemForURL(url);
+        continue;
       }
-      // Remove blocked item from existing users data.
-      if (!IsBlockedBuiltInItem(built_in_item))
-        items_.push_back(built_in_item);
-      continue;
-    }
-
-    bool open_in_panel;
-    if (const auto value = item.FindBoolKey(kSidebarItemOpenInPanelKey)) {
-      open_in_panel = *value;
-    } else {
-      continue;
+      // Always use latest properties for built-in type item.
+      if (type == SidebarItem::Type::kTypeBuiltIn) {
+        if (const auto value =
+                item.FindIntKey(kSidebarItemBuiltInItemTypeKey)) {
+          auto id = static_cast<SidebarItem::BuiltInItemType>(*value);
+          auto iter = std::find_if(
+              default_items_to_add.begin(), default_items_to_add.end(),
+              [id](const auto& default_item) {
+                return default_item.built_in_item_type == id;
+              });
+          // It might be an item which is no longer is offered as built-in
+          if (iter == default_items_to_add.end()) {
+            continue;
+          }
+          // Valid built-in item, add it
+          items_.emplace_back(*std::make_move_iterator(iter));
```

Comment:
I think moving this way looks really great but If we apply NoDestructor() and const& to GetDefaultSidebarItems(), this may not work 😥

Reply:
I don't see a suggestion to apply const& to `GetDefaultSidebarItems()`, just `GetDefaultBuiltInItemTypes()` ;-) So for now we are ok to keep this style, and we can debate if `GetDefaultSidebarItems()` should be const another time.

Reply:
Oh, yes. I was mistaken. Thanks!